static/js/narrow.js

```diff
@@ -845,6 +847,7 @@ function handle_post_narrow_deactivate_processes() {
 }
 
 export function deactivate(coming_from_recent_topics = false) {
+    has_shown_message_list_view = true;
```

Comment:
Will this be incorrectly `false` if you load directly to "All messages"?

Reply:
It shouldn't, because `deactivate` sets it as `true` based on the following path:

deactivate | @ | narrow.js:852
show_all_message_view | @ | hashchange.js:92
do_hashchange_normal | @ | hashchange.js:173
hashchanged | @ | hashchange.js:364
initialize | @ | hashchange.js:373
initialize_everything | @ | ui_init.js:659
(anonymous) | @ | ui_init.js:685
measure_time | @ | blueslip.ts:239
(anonymous) | @ | ui_init.js:684

Reply:
Yes, but if you load the app with all messages as the initial/default view, I don't think `deactivated` is ever called?

Reply:
If I add
<details><summary> a console.trace()</summary>

```diff
diff --git a/static/js/narrow.js b/static/js/narrow.js
index a969f4f03b..e03f046ebf 100644
--- a/static/js/narrow.js
+++ b/static/js/narrow.js
@@ -847,6 +847,7 @@ function handle_post_narrow_deactivate_processes() {
 }
 
 export function deactivate(coming_from_recent_topics = false) {
+    console.trace();
     has_shown_message_list_view = true;
     // NOTE: Never call this function independently,
     // always use browser_history.go_to_location("#all_messages") to

```
</details>

If I leave the default view to be recent topics, and open a new tab, open the console, and then navigate to `localhost:9991/#all_messages`:

<details><summary>here's the complete log I get</summary>

deactivate | @ | narrow.js:850
-- | -- | --
show_all_message_view | @ | hashchange.js:94
do_hashchange_normal | @ | hashchange.js:186
hashchanged | @ | hashchange.js:377
initialize | @ | hashchange.js:386
initialize_everything | @ | ui_init.js:661
(anonymous) | @ | ui_init.js:683
measure_time | @ | blueslip.ts:239
(anonymous) | @ | ui_init.js:682
mightThrow | @ | jquery.js:3766
process | @ | jquery.js:3834
setTimeout (async) |   |  
(anonymous) | @ | jquery.js:3872
fire | @ | jquery.js:3500
fireWith | @ | jquery.js:3630
fire | @ | jquery.js:3638
fire | @ | jquery.js:3500
fireWith | @ | jquery.js:3630
ready | @ | jquery.js:4110
setTimeout (async) |   |  
(anonymous) | @ | jquery.js:4131
(anonymous) | @ | jquery.js:28
../../../srv/zulip-npm-cache/dfc230fa976bd1b456ba23dac4ec14c83da28611/node_modules/jquery/dist/jquery.js | @ | jquery.js:40
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
../../../srv/zulip-npm-cache/dfc230fa976bd1b456ba23dac4ec14c83da28611/node_modules/jquery/dist/jquery-exposed.js | @ | jquery-exposed.js:1
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/page_params.ts | @ | i18n.ts:138
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/webpack_public_path.js | @ | webpack_public_path.js:7
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/bundles/common.js | @ | blueslip_stacktrace.ts:84
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/bundles/app.js | @ | jquery.idle.js:121
__webpack_require__ | @ | bootstrap:24
(anonymous) | @ | startup:6
__webpack_require__.O | @ | chunk loaded:25
(anonymous) | @ | startup:8
(anonymous) | @ | startup:8
</details>

If set the default view to be all messages, open a new tab, open the console, and then navigate to `localhost:9991`:

<details><summary>here's the complete log I get</summary>

deactivate | @ | narrow.js:850
-- | -- | --
show_all_message_view | @ | hashchange.js:94
show_default_view | @ | hashchange.js:116
do_hashchange_normal | @ | hashchange.js:180
hashchanged | @ | hashchange.js:377
initialize | @ | hashchange.js:386
initialize_everything | @ | ui_init.js:661
(anonymous) | @ | ui_init.js:683
measure_time | @ | blueslip.ts:239
(anonymous) | @ | ui_init.js:682
mightThrow | @ | jquery.js:3766
process | @ | jquery.js:3834
setTimeout (async) |   |  
(anonymous) | @ | jquery.js:3872
fire | @ | jquery.js:3500
fireWith | @ | jquery.js:3630
fire | @ | jquery.js:3638
fire | @ | jquery.js:3500
fireWith | @ | jquery.js:3630
ready | @ | jquery.js:4110
setTimeout (async) |   |  
(anonymous) | @ | jquery.js:4131
(anonymous) | @ | jquery.js:28
../../../srv/zulip-npm-cache/dfc230fa976bd1b456ba23dac4ec14c83da28611/node_modules/jquery/dist/jquery.js | @ | jquery.js:40
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
../../../srv/zulip-npm-cache/dfc230fa976bd1b456ba23dac4ec14c83da28611/node_modules/jquery/dist/jquery-exposed.js | @ | jquery-exposed.js:1
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/page_params.ts | @ | i18n.ts:138
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/webpack_public_path.js | @ | webpack_public_path.js:7
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/bundles/common.js | @ | blueslip_stacktrace.ts:84
__webpack_require__ | @ | bootstrap:24
fn | @ | hot module replacement:61
./static/js/bundles/app.js | @ | jquery.idle.js:121
__webpack_require__ | @ | bootstrap:24
(anonymous) | @ | startup:6
__webpack_require__.O | @ | chunk loaded:25
(anonymous) | @ | startup:8
(anonymous) | @ | startup:8
</details>

Because I see those logs, in both cases I expect `has_shown_message_list_view` to be set correctly.
I think the important path to notice from those trace backs is that `ui_init` calls `hashchange.initialize();` which eventually calls `show_all_message_view` in both cases, and `show_all_message_view` calls `deactivate`

Am I missing anything?

Reply:
OK, great, thanks for confirming; I hadn't traced the code path and was just working off recollection. I merged this commit as c4bb181056ab0ac40c9340f460039bc4cda502f0, after moving the state logging to a more appropriate place (after all the places we might return without doing anything).

It'd probably have been better to skip the tracebacks in your comment, BTW; I think just stating "ui_init calls hashchange.initialize(); which eventually calls show_all_message_view in both cases, and show_all_message_view calls deactivate" would have been easier to read :).